src/net_processing.cpp

```diff
@@ -220,18 +223,68 @@ struct CNodeState {
     }
 };
 
-/** Map maintaining per-node state. Requires cs_main. */
-map<NodeId, CNodeState> mapNodeState;
+/** Provides locked access to a CNodeState */
+class CNodeStateAccessor {
+private:
+    std::shared_ptr<CNodeState> pstate;
 
-// Requires cs_main.
-CNodeState *State(NodeId pnode) {
-    map<NodeId, CNodeState>::iterator it = mapNodeState.find(pnode);
-    if (it == mapNodeState.end())
-        return NULL;
-    return &it->second;
+    CNodeStateAccessor() : pstate() { }
+    CNodeStateAccessor(std::shared_ptr<CNodeState>&& pstateIn) : pstate(std::move(pstateIn)) { ENTER_CRITICAL_SECTION(pstate->cs); }
+    friend class NodeStateStorage;
+
+public:
+    ~CNodeStateAccessor() { if (pstate) LEAVE_CRITICAL_SECTION(pstate->cs); }
+
+    CNodeStateAccessor(const CNodeStateAccessor&) =delete;
+    CNodeStateAccessor& operator= (const CNodeStateAccessor&) =delete;
+
+    CNodeStateAccessor(CNodeStateAccessor&& o) { pstate = o.pstate; o.pstate = NULL; }
+
+    bool operator()() const { return (bool)pstate; }
```

Comment:
I dunno I dont really like the bool operator. If you really prefer it I'm happy to change it.